apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,247 @@
+        /// </summary>
+        /// <param name="message">The message if the condition is violated.</param>
+        /// <param name="controlling">The value controlling whether or not any other value can be non-null.</param>
+        /// <param name="values">The values checked for non-nullity if <paramref name="controlling"/> is non-null.</param>
+        protected void ValidateOptionExcludesOthers(string message, object controlling, params object[] values)
+        {
+            GaxPreconditions.CheckState(controlling == null || !values.Contains(null), message);
+        }
+
+        /// <summary>
+        /// Creates a call invoker synchronously. Override this method in a concrete builder type if more
+        /// call invoker mechanisms are supported.
+        /// </summary>
+        protected virtual CallInvoker CreateCallInvoker()
+        {
+            if (CallInvoker != null)
+            {
+                return CallInvoker;
+            }
+            var endpoint = Endpoint ?? GetDefaultEndpoint();
+            var credentials = GetChannelCredentials();
+            Channel channel = new Channel(endpoint.Host, endpoint.Port, GetChannelCredentials());
+            return new DefaultCallInvoker(channel);
+        }
+
+        /// <summary>
+        /// Creates a call invoker asynchronously. Override this method in a concrete builder type if more
+        /// call invoker mechanisms are supported.
+        /// </summary>
+        protected virtual async Task<CallInvoker> CreateCallInvokerAsync(CancellationToken cancellationToken)
+        {
+            if (CallInvoker != null)
+            {
+                return CallInvoker;
+            }
+            var endpoint = Endpoint ?? GetDefaultEndpoint();
+            var credentials = await GetChannelCredentialsAsync(cancellationToken).ConfigureAwait(false);
+            Channel channel = new Channel(endpoint.Host, endpoint.Port, credentials);
+            return new DefaultCallInvoker(channel);
+        }
+
+        /// <summary>
+        /// Obtains channel credentials synchronously. Override this method in a concrete builder type if more
+        /// credential mechanisms are supported.
+        /// </summary>
+        protected virtual ChannelCredentials GetChannelCredentials()
+        {
+            if (ChannelCredentials != null)
+            {
+                return ChannelCredentials;
+            }
+            GoogleCredential unscoped;
+            if (CredentialsPath != null)
+            {
+                unscoped = GoogleCredential.FromFile(CredentialsPath);
+            }
+            else if (JsonCredentials != null)
+            {
+                unscoped = GoogleCredential.FromJson(JsonCredentials);
+            }
+            else
+            {
+                unscoped = GoogleCredential.GetApplicationDefault();
+            }
+            return unscoped.CreateScoped(Scopes ?? GetDefaultScopes()).ToChannelCredentials();
```

Comment:
If CreateScoped does not clone or copy Scopes in some way, this might be the time to turn it into a readonly list.

Reply:
Good point - it looks like it really doesn't do that. We don't particularly need it to be read-only, but we should create a one-time copy.